Add option to hard clip read in CigarUtils (replacement for #1453)

kachulis opened the pull request:
@lbergelson here is the replacement PR for #1453 we discussed earlier today.

@fleharty, @lbergelson and I talked about how to get hard clipping merged ASAP, given that I do not have permissions to push directly to htsjdk, and so cannot make modifications directly to your branch.  He suggested opening up a new PR to replace the old one, which is what this PR is.

lbergelson commented on src/main/java/htsjdk/samtools/util/CigarUtil.java:
```diff
@@ -180,11 +180,14 @@ public static void clip3PrimeEndOfRead(SAMRecord rec, final int clipFrom, final
         if(clippingOperator == CigarOperator.HARD_CLIP) {
-            byte[] bases = rec.getReadBases();
+            final byte[] bases = rec.getReadBases();
+            final byte[] baseQualities = rec.getBaseQualities();
```
lets make sure bases.length and baseQualities.length are the same, and then use the same values for the start/stop for each.

kachulis replied:
done

kachulis commented on src/test/java/htsjdk/samtools/util/CigarUtilTest.java:
```diff
@@ -174,8 +173,51 @@ public void addingSoftClippedBasesTest(final String testName, final String cigar
+                                 final int clipFrom, final String expectedReadString, final String expectedCigar) throws IOException {
+
+        final SAMRecord rec = createTestSamRec(initialReadString, initialCigar, negativeStrand);
+//        Assert.assertEquals(rec.getCigarString(), initialCigar, testName);
```
was already removed in commit 696541fc45774030e033aa849000abda938e868c

kachulis commented on src/main/java/htsjdk/samtools/util/CigarUtil.java:
```diff
@@ -89,25 +91,40 @@
+     * @param clippingOperator clipping operator to be merged
+     * @param trailingHardClippedBases number of hardClippedBases which were on the end of the original cigar
+     */
+    static private void mergeClippingCigarElement(List<CigarElement> newCigar, CigarElement c,
```
done

lbergelson commented on src/main/java/htsjdk/samtools/util/CigarUtil.java:
```diff
@@ -201,6 +212,13 @@ public static void clip3PrimeEndOfRead(SAMRecord rec, final int clipFrom, final
             }
         }
 
+        if (rec.getReadLength() != originalReadLength) {
```
Could you add a method comment saying that this will remove these tags?

kachulis replied:
done

lbergelson commented on src/main/java/htsjdk/samtools/util/CigarUtil.java:
```diff
@@ -201,6 +212,13 @@ public static void clip3PrimeEndOfRead(SAMRecord rec, final int clipFrom, final
 
+        if (rec.getReadLength() != originalReadLength) {
+            //invalidate NM, UQ, MD tags if we have changed the length of the read.
+            rec.setAttribute(SAMTag.NM.name(), null);
```
Add a test that shows these get invalidated.

kachulis replied:
done

kachulis commented:
@lbergelson I added documentation and tests for the tag invalidations.  Also adjusted the logic to only invalidate in the number of reference bases the read aligns to changes (no need to invalidate these tags if we are harclipping bases that were already softclipped).

I don't think I have merge permissions, so I'm happy with you merging this at this point if you approve.
